### ISSUE TITLE
MSL: atomic writes to Texture3D issues

### ISSUE THREAD

**Issue by jmccarthy634** [KhronosGroup/SPIRV-Cross#1537]
Hi, I've noticed a couple more issues with the workaround for atomic writes to textures in MSL - the first is that with argument buffers enabled for the following HLSL:

```
RWTexture3D<int> g_gridTexture : register(u0);

[numthreads(256, 1, 1)]
void gridUpdate(uint3 GTid : SV_GroupThreadID,
	uint3 gId : SV_GroupID,
	uint gIdx : SV_GroupIndex,
	uint3 dID : SV_DispatchThreadID)
{
    InterlockedAdd(g_gridTexture[gId], gIdx);
}
```

generates this MSL:

```
#pragma clang diagnostic ignored "-Wunused-variable"

#include <metal_stdlib>
#include <simd/simd.h>
#include <metal_atomic>

using namespace metal;

struct spvDescriptorSetBuffer0
{
    texture3d<int> g_gridTexture [[id(0)]];
    device atomic_uint* g_gridTexture_atomic [[id(1)]];
};

kernel void gridUpdate(constant spvDescriptorSetBuffer0& spvDescriptorSet0 [[buffer(0)]], uint3 gl_WorkGroupID [[threadgroup_position_in_grid]], uint gl_LocalInvocationIndex [[thread_index_in_threadgroup]])
{
    int _24 = atomic_fetch_add_explicit((device atomic_int*)&spvDescriptorSet0.g_gridTexture_atomic[gl_WorkGroupID], int(gl_LocalInvocationIndex), memory_order_relaxed);
}
```

The texture3d<int> is shadowed by an atomic_uint*, which should obviously be an atomic_int* instead. The other more serious issue here is that it attempts to access that atomic_uint using a uint3, which is obviously not going to compile. This is also problematic because [it's not possible to alias a 3D texture from a buffer](https://developer.apple.com/documentation/metal/mtlbuffer/1613852-newtexturewithdescriptor?language=objc) so there may not be too much that can be done about this anyway. Regardless, here is the source HLSL and the generated spirv and MSL: 
[Texture3D_atomic.zip](https://github.com/KhronosGroup/SPIRV-Cross/files/5520898/Texture3D_atomic.zip)

**Comment by cdavis5e** [comment]
> The texture3d is shadowed by an atomic_uint*, which should obviously be an atomic_int* instead.

If you notice, it cast the pointer to `device atomic_int*` in the actual atomic call. The real problem here is that it's taking the address of the pointer itself, instead of passing the pointer directly. This really needs to be addressed.

> The other more serious issue here is that it attempts to access that atomic_uint using an int3, which is obviously not going to compile.

This is because we don't handle the `texture3d` case, at all, precisely because, as you yourself mentioned:

> This is also problematic because [it's not possible to alias a 3D texture from a buffer](https://developer.apple.com/documentation/metal/mtlbuffer/1613852-newtexturewithdescriptor?language=objc)

For this case, we'd need a function which accepts the texel coordinates and returns the corresponding index into the buffer, which is how 2D textures are handled.

**Comment by jmccarthy634** [comment]
Would it be possible/worth having some kind of explicit error in the MSL compiler for the specific case of trying to use atomic operations with 3D textures? Because yeah, regardless of how we access into the buffer it's just not going to be possible to create that underlying buffer in the first place.

**Comment by cdavis5e** [comment]
> Would it be possible/worth having some kind of explicit error in the MSL compiler for the specific case of trying to use atomic operations with 3D textures?

Yeah, we probably ought to be throwing an error when trying to use atomics on an image that is not an `image2D` or a `imageBuffer`, or an `image1D` if we're treating 1D as 2D.
